retrybp/filters.go

```diff
@@ -0,0 +1,160 @@
+package retrybp
+
+import (
+	"context"
+	"errors"
+	"net"
+
+	retry "github.com/avast/retry-go"
+
+	"github.com/reddit/baseplate.go/clientpool"
+)
+
+var (
+	// DefaultRetryDecision builds the default decision used by Filters.
+	//
+	// If none of the RetryFilters make a decision on whether or not to retry the
+	// request, the "Retry" field of the default decision will be returned.
+	DefaultRetryDecision = DoNotRetry
+)
+
+// Filters returns a `retry.RetryIf` function that checks the error against the
+// given filters and returns either the decision reached by a filter or the
+// DefaultRetryDecision.
+//
+// You should not use this with any other retry.RetryIf options as one will
+// override the other.
+func Filters(filters ...RetryFilter) retry.Option {
+	return retry.RetryIf(
+		func(err error) bool {
+			for _, filter := range filters {
+				decision := filter(err)
+				if decision.Reached {
+					return decision.Retry
+				}
+			}
+			return DefaultRetryDecision().Retry
+		},
+	)
+}
+
+// FilterDecision is returned by a RetryFilter to communicate whether it has
+// decided whether or not a request should be retried.
+//
+// Generally, the default decision for a RetryFilter should be "NoDecision".  A
+// RetryFilter should check a single case to determine one of "DoSend" or
+// "DoNotSend" and should return "NoDecision" if that check fails.
+type FilterDecision struct {
```

Comment:
Any particular reason for this Decision struct over the approach I've seen more typically, which is to have a FilterChain and each filter has the choice on whether to call the next one?

I guess I'm saying that a filter "linked-list" is more common choice than a filter "array".  The "linked-list" might provide a little more flexibility and not require secondary structs.

Don't hate me for quoting the Servlet API :) https://tomcat.apache.org/tomcat-5.5-doc/servletapi/javax/servlet/Filter.html

Reply:
This is just an approach I'm familiar with, but the function chain makes sense as well.

I'm open to changing it unless others feel strongly about one way or the other.

@fishy / @konradreiche ?

Reply:
I think both have their pros and cons, the chain approach seems a little more... elegant?

The nice thing about the current approach though is the return struct forces you to say "Done" vs "Not Done" while the chain approach means that any custom filters have to remember to call `chain.Next()`.

Reply:
I like the chain design more than a struct with two bools :)

Reply:
That works for me :)

OOC, I'm going to benchmark both and see what it looks like (probably pretty much the same).

Reply:
Same